# Capstone Project: The Battle of the Neighborhoods
### Applied Data Science Capstone by IBM/Coursera

## Table of contents
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussion](#results)
* [Conclusion](#conclusion)

## Introduction: Business Problem <a name="introduction"></a>

In this project we will try to find out an optimal livable location in a city.
By livable location here it means the number of facilities, nearby venues provided in the vicinity of a location. This report specifically targets stakeholders who are interested in finding a livable location in **Pune city, India**.

Also, the stakeholder would choose a particular location which is closest to its workplace. Hence, we would categorize various locations in the city on the basis of their distance between **the most prominent workplaces/ industrial zones in the city** and **nearby venues to those locations**

We would cluster all the important/livable and recommended regions where someone can live which have the most number of amenities/venues in the locality. User/Stakeholder can choose according the location nearest to his/her office/campus. 

For example: - Hinjewadi and Magarpatta are two IT company zones which are situated in opposite direction to each other, If someone lives in a location near to Hinjewadi and his/her company moves to Magarpatta, which location should be suggested having similar facilities/ venues compared to his/her current living location.

We would identify areas with most promising characteristics and their advantages will then be clearly expressed, so that best possible final location may be chosen by our stakeholders.

## Data <a name="data"></a>

We have taken latitude and longitude of most prominent pune locations from ___[PMC Open Data Store](http://opendata.punecorporation.org/Citizen/CitizenDatasets/Index)___ offical website and __[Kaggle Dataset](https://www.kaggle.com/dynamic22/pune-property-prices)__. The data derived from kaggle dataset do not have latitude longitude information, hence we have used geopy library to fetch latitude and longitude values for such locations.

In [1]:
#Let's try to read dataset
import pandas as pd
location_df=pd.read_excel('dataset.xlsx', sheet_name='Sheet2',  header=0, nrows=199)


In [2]:
import requests
location_df

Location   Latitude  Longitude
0                      Bund Garden  18.539848  73.885117
1                     Shivajinagar  18.510099  73.817398
2                            Aundh  18.563162  73.809555
3                          Kondhwa  18.478436  73.890213
4                        Chinchwad  18.636131  73.796143
5                      Satara Road  18.488499  73.857956
6                          Kothrud  18.508699  73.812500
7              Senapati Bapat Road  18.534451  73.837349
8                    Kalyani Nagar  18.548101  73.900070
9                 Hinjewadi Phase1  18.586555  73.734741
10                Hinjewadi Phase2  18.598255  73.706207
11                 Magarpatta City  18.522141  73.931740
12                      VimanNagar  18.552900  73.892197
13                           Baner  18.559280  73.790627
14                       Hinjewadi  18.592600  73.718399
15                          Kirkee  18.573057  73.879478
16                    Fatima Nagar  18.503725  73.900597
17                          Pimpri  18.622713  73.802017
18    Model Colony - Wealth Branch  18.530758  73.835197
19                 Pimple Saudagar  18.596691  73.797791
20                   Sinhagad Road  18.472366  73.821587
21                      Tilak Road  18.507809  73.844376
22                         Bavdhan  18.520041  73.778526
23                          Katraj  18.447878  73.857796
24           Aundh - Nagardas Road  18.565594  73.806984
25   Koregaon Park - Wealth Branch  18.538630  73.885826
26                         Kharadi  18.520477  73.856956
27                           Nigdi  18.653700  73.770699
28                   Thermax Chowk  18.660992  73.794052
29                    Mayur Colony  18.505028  73.817574
..                             ...        ...        ...
168                   Sasane Nagar  18.499571  73.933818
169                   Shanti Nagar  19.064192  72.931151
170                  Shastri Nagar  19.138039  72.828016
171                      Shikrapur  18.695989  74.136901
172                   Shinde Vasti  18.649108  74.193584
173                  Shivaji Nagar  19.063789  72.924900
174                        Shivane  18.467041  73.793065
175                   Siddhi Nagar  18.496901  73.823339
176                  Someshwarwadi  18.549381  73.798292
177                     Sopan Baug  18.514171  73.901401
178                            Sus  18.554159  73.753904
179                      Sutarwadi  18.538629  73.781421
180                      Tathawade  18.493309  73.827318
181                       TherGaon  18.609275  73.772937
182                    Tukai Nagar  18.473799  73.827719
183                  Tukaram Nagar  18.614590  73.815697
184                    Udyog Nagar  19.170000  72.846000
185                          Undri  18.458157  73.915029
186                 VadGaon Budruk  18.467497  73.825365
187                  Vallabh Nagar  19.243475  72.866084
188                    Viman Nagar  18.520306  73.782938
189                   Vishal Nagar  18.585322  73.782938
190                  Vitthal Nagar  18.448006  73.814689
191                    Vivek Nagar  19.076176  72.905625
192                 WadGaon Budruk  18.460143  73.826010
193                  WadGaon Sheri  18.550441  73.917167
194                          Wakad  18.610902  73.763796
195                       Wanowrie  18.488368  73.898667
196                 Yashwant Nagar  19.077216  72.854130
197                        Yerwada  18.550361  73.879746

[198 rows x 3 columns]

Let's import folium and locate this points on Pune map

In [3]:
import folium
from geopy.geocoders import Nominatim
print('libraries imported')

libraries imported


#### Use geopy library to get longitude and latitude of Pune city, India

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>pune_explorer</em>, as shown below.

The below code is used to fetch and store latitude and longitude values in excel file.

In [4]:
import math
latlon = []
for index in range(66,121):
    if math.isnan(location_df.loc[index,'Latitude']):
        address = location_df.loc[index,'Location'] + ', Pune IN'
        geolocator = Nominatim(user_agent="pune_explorer")
        location = geolocator.geocode(address)
        if location is not None:
            location_df.loc[index,'Latitude'] = location.latitude
            location_df.loc[index,'Longitude'] = location.longitude
            print('The geograpical coordinate of {} are {}, {}.'.format(address, latitude, longitude))
location_df.to_excel('dataset1.xlsx')

We will also use location of 10 industrial areas in Pune. This data is also taken from ___[PMC Open Data Store](http://opendata.punecorporation.org/Citizen/CitizenDatasets/Index)___ offical website. Let's read that file

In [5]:
industries_df=pd.read_excel('Major Industries.xlsx', sheet_name='Sheet1',  header=0, nrows=10)
industries_df

Industries   Latitude  Longitude
0                              Pimpri Chinchwad MIDC  18.627929  73.800983
1       Rajiv Gandhi InfoTech Park Hinjewadi Phase I  18.591684  73.734782
2      Rajiv Gandhi InfoTech Park Hinjewadi Phase II  18.598255  73.706207
3  Rajiv Gandhi InfoTech Park Hinjewadi Phase III...  18.591770  73.733895
4                                    Magarpatta City  18.522141  73.931740
5                             Kharadi Knowledge Park  18.550518  73.942494
6                             Talawade InfoTech Park  18.739658  73.806857
7                         Talegaon Floriculture Park  18.729488  73.654067
8                         Ranjangaon Industrial Area  18.753635  74.244579
9                             Chakan Industrial Area  18.762311  73.862545

Now, create a map of Pune city with nearby locations superimposed on top

In [6]:
# create map of Pune using latitude and longitude values
address = 'Pune, IN'
geolocator = Nominatim(user_agent="pune_explorer")
location = geolocator.geocode(address)
latitude=location.latitude
longitude=location.longitude
map_pune = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, Location in zip(location_df['Latitude'], location_df['Longitude'], location_df['Location']):
    label = Location
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_pune)  
    
map_pune

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [7]:
CLIENT_ID = 'TR4OJXMM340Z5YTVDZ1QGAG1YHPBOJ4NEPK4V52K10Y0RYPY' # your Foursquare ID
CLIENT_SECRET = 'DZFZHGTHTBEWNOM25GNI24LBCGGUTLVPQMALSZQBQQWSIZMT' # your Foursquare Secret
VERSION = '20190324' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: TR4OJXMM340Z5YTVDZ1QGAG1YHPBOJ4NEPK4V52K10Y0RYPY
CLIENT_SECRET:DZFZHGTHTBEWNOM25GNI24LBCGGUTLVPQMALSZQBQQWSIZMT


## Explore Neighborhoods in Pune

Let's borrow the **get_category_type** function from the Foursquare lab.

In [8]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

#### Let's create a function to retrieve nearby venues in Pune

We will use 1 Km as radius this time

In [9]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    LIMIT=100
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now lets write the code to run the above function on each neighborhood and create a new dataframe.

In [10]:
pune_venues = getNearbyVenues(names=location_df['Location'],
                                   latitudes=location_df['Latitude'],
                                   longitudes=location_df['Longitude']
                                  )

Bund Garden
Shivajinagar
Aundh
Kondhwa
Chinchwad
Satara Road
Kothrud
Senapati Bapat Road
Kalyani Nagar
Hinjewadi Phase1
Hinjewadi Phase2
Magarpatta City
VimanNagar
Baner
Hinjewadi
Kirkee
Fatima Nagar
Pimpri
Model Colony - Wealth Branch
Pimple Saudagar
Sinhagad Road
Tilak Road
Bavdhan
Katraj
Aundh - Nagardas Road
Koregaon Park - Wealth Branch
Kharadi
Nigdi
Thermax Chowk
Mayur Colony
Sus Pashan Road
WTC-Kharadi
Navi Peth
Warje
Murti - Baramati
Karanjepul
Kamthadi
Kikvi
Malad Patas
Deulgaon Raje
Pimpalgaon - Daund
Pargaon
B T Kawade
Balewadi, Maharashtra
Karve Nagar
Nanded City, Maharashtra
Bhigwan
Sahakar Nagar
Bhandarkar Road
Raviwar Peth
Sadashiv Peth
Erandavana
Camp
Paud Road
Ghole Road
Blueridge Hinjewadi
Ravet
New Sanghvi
Pirangut
Narhe
Wagholi
Mohammadwadi
Bhosle Nagar
Undri Pisoli
E Square University Road
Vishrantwadi
Nana Peth
Fursungi
Salunke Vihar Road
Manjari Road Hadapsar
Phoenix Mall
Shivar Garden Chowk
Baner-D-Mart Complex
Null Stop-Karve Road
Market Yard
Pimple Nilakh
Prab

In [11]:
print(pune_venues.shape)
pune_venues.head()

(5431, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Bund Garden              18.539848               73.885117   
1  Bund Garden              18.539848               73.885117   
2  Bund Garden              18.539848               73.885117   
3  Bund Garden              18.539848               73.885117   
4  Bund Garden              18.539848               73.885117   

                               Venue  Venue Latitude  Venue Longitude  \
0                        La Pizzeria       18.539621        73.883401   
1         Hidden Place - The Hangout       18.539651        73.887023   
2                         Savya Rasa       18.538874        73.886561   
3  Starbucks Coffee: A Tata Alliance       18.539341        73.886602   
4                       Little Italy       18.539598        73.883464   

            Venue Category  
0       Italian Restaurant  
1                      Pub  
2  South Indian Restaurant  
3              Coffee Shop  
4       Italian Restaurant

Let's check how many venues were returned for each neighborhood

In [12]:
pune_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                   
Adarsh Nagar                              100                     100    100   
Akurdi                                     17                      17     17   
Alandi                                      5                       5      5   
Alandi Road                                 5                       5      5   
Ambedkar Nagar                            100                     100    100   
Anand Nagar                                12                      12     12   
Anand Park Nagar                           15                      15     15   
Ashok Nagar                                11                      11     11   
Aundh                                      59                      59     59   
Aundh - Nagardas Road                      51                      51     51   
Aundh Annexe                               78                      78     78   
Aundh Gaon                                 65                      65     65   
Azad Nagar                                 29                      29     29   
B T Kawade                                 12                      12     12   
Balewadi Phata                             30                      30     30   
Balewadi, Maharashtra                      53                      53     53   
Baner                                      55                      55     55   
Baner Pashan Link Road                     50                      50     50   
Baner-D-Mart Complex                       41                      41     41   
Baner-Sus                                  46                      46     46   
Bavdhan                                    18                      18     18   
Bhandarkar Road                            85                      85     85   
Bharati Vidyapeeth                         16                      16     16   
Bhigwan                                     5                       5      5   
Bhosale Nagar                              16                      16     16   
Bhosari                                     4                       4      4   
Bhosle Nagar                               19                      19     19   
Bhusari Colony                             12                      12     12   
Bibwewadi                                  13                      13     13   
Blueridge Hinjewadi                        14                      14     14   
...                                       ...                     ...    ...   
Sopan Baug                                  9                       9      9   
Sus                                         2                       2      2   
Sus Pashan Road                            15                      15     15   
Sutarwadi                                  13                      13     13   
Tathawade                                  24                      24     24   
TherGaon                                    6                       6      6   
Thermax Chowk                               5                       5      5   
Tilak Road                                 46                      46     46   
Tukai Nagar                                15                      15     15   
Tukaram Nagar                              11                      11     11   
Udyog Nagar                                33                      33     33   
Undri                                       5                       5      5   
Undri Pisoli                                2                       2      2   
VadGaon Budruk                             15                      15     15   
Vallabh Nagar                              14                      14     14   
Viman Nagar                                16                      16     16   
VimanNagar                                 23                      23     23   
Vishal Nagar            

#### Let's find out how many unique categories can be curated from all the returned venues

In [13]:
print('There are {} uniques categories.'.format(len(pune_venues['Venue Category'].unique())))

There are 233 uniques categories.


I will use these 3 datasets to determine similarity across various locations in Pune city.

## Methodology <a name="methodology"></a>

We are going to analyze our datasets in two different cases: -

1.) Top 10 venues at each location will be calculated, then we will run K-Means clustering algorithm to determine common patterns in these locations.

2.) We will calculate the distance of each location from the ten industrial areas used in our analysis. And we will determine the top 10 venues at each location and then, we will run K-Means clustering algorithm to determine common patterns when distance from industrial areas is also considered.

#### Let's draw a scatter plot of all the industrial areas in Pune location

In [14]:
# create map of Pune containg industries superimposed on top using their latitude and longitude values
address = 'Pune, IN'
geolocator = Nominatim(user_agent="pune_explorer")
location = geolocator.geocode(address)
latitude=location.latitude
longitude=location.longitude
map_pune_industries = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, Location in zip(industries_df['Latitude'], industries_df['Longitude'], industries_df['Industries']):
    label = Location
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_pune_industries)  
    
map_pune_industries

## Analyze Each Neighborhood <a name="analysis"></a>

We will analyze all the venues across various neighborhood

In [15]:
# one hot encoding
pune_onehot = pd.get_dummies(pune_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
pune_onehot['Neighborhood'] = pune_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [pune_onehot.columns[-1]] + list(pune_onehot.columns[:-1])
pune_onehot = pune_onehot[fixed_columns]

pune_onehot.head()

Neighborhood  ATM  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Bund Garden    0                  0               0                  0   
1  Bund Garden    0                  0               0                  0   
2  Bund Garden    0                  0               0                  0   
3  Bund Garden    0                  0               0                  0   
4  Bund Garden    0                  0               0                  0   

   Airport  Airport Lounge  American Restaurant  Andhra Restaurant  Arcade  \
0        0               0                    0                  0       0   
1        0               0                    0                  0       0   
2        0               0                    0                  0       0   
3        0               0                    0                  0       0   
4        0               0                    0                  0       0   

  ...   Track Stadium  Trail  Train Station  Tunnel  \
0 ...               0      0              0       0   
1 ...               0      0              0       0   
2 ...               0      0              0       0   
3 ...               0      0              0       0   
4 ...               0      0              0       0   

   Vegetarian / Vegan Restaurant  Warehouse Store  Watch Shop  Women's Store  \
0                              0                0           0              0   
1                              0                0           0              0   
2                              0                0           0              0   
3                              0                0           0              0   
4                              0                0           0              0   

   Yoga Studio  Zoo  
0            0    0  
1            0    0  
2            0    0  
3            0    0  
4            0    0  

[5 rows x 234 columns]

And let's examine the new dataframe size.

In [16]:
pune_onehot.shape

(5431, 234)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [17]:
pune_grouped=pune_onehot.groupby('Neighborhood').mean().reset_index()
pune_grouped.head()

Neighborhood  ATM  Accessories Store  Adult Boutique  Afghan Restaurant  \
0    Adarsh Nagar  0.0                0.0             0.0                0.0   
1          Akurdi  0.0                0.0             0.0                0.0   
2          Alandi  0.0                0.0             0.0                0.0   
3     Alandi Road  0.0                0.0             0.0                0.0   
4  Ambedkar Nagar  0.0                0.0             0.0                0.0   

   Airport  Airport Lounge  American Restaurant  Andhra Restaurant  Arcade  \
0      0.0             0.0                 0.01                0.0    0.00   
1      0.0             0.0                 0.00                0.0    0.00   
2      0.0             0.0                 0.00                0.0    0.00   
3      0.0             0.0                 0.00                0.0    0.00   
4      0.0             0.0                 0.01                0.0    0.02   

  ...   Track Stadium  Trail  Train Station  Tunnel  \
0 ...             0.0    0.0            0.0     0.0   
1 ...             0.0    0.0            0.0     0.0   
2 ...             0.0    0.0            0.0     0.0   
3 ...             0.0    0.0            0.0     0.0   
4 ...             0.0    0.0            0.0     0.0   

   Vegetarian / Vegan Restaurant  Warehouse Store  Watch Shop  Women's Store  \
0                           0.01              0.0         0.0           0.01   
1                           0.00              0.0         0.0           0.00   
2                           0.00              0.0         0.0           0.00   
3                           0.00              0.0         0.0           0.00   
4                           0.00              0.0         0.0           0.00   

   Yoga Studio  Zoo  
0          0.0  0.0  
1          0.0  0.0  
2          0.0  0.0  
3          0.0  0.0  
4          0.0  0.0  

[5 rows x 234 columns]

#### Let's confirm the new size

In [18]:
pune_grouped.shape

(188, 234)

#### Let's write a function to sort the venues in descending order.

In [19]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [20]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = pune_grouped['Neighborhood']

for ind in np.arange(pune_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(pune_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0    Adarsh Nagar     Indian Restaurant                  Café   
1          Akurdi                   Gym                  Café   
2          Alandi  Fast Food Restaurant           Bus Station   
3     Alandi Road                Bakery           Supermarket   
4  Ambedkar Nagar     Indian Restaurant                   Bar   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0    Chinese Restaurant                   Pub                Bakery   
1           Coffee Shop     Convenience Store        Ice Cream Shop   
2     Indian Restaurant                 River            Distillery   
3             Juice Bar           Coffee Shop                 Hotel   
4    Italian Restaurant           Coffee Shop                Lounge   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0           Coffee Shop  Gym / Fitness Center             Asian Restaurant   
1           Snack Place  Fast Food Restaurant    Middle Eastern Restaurant   
2        Farmers Market                  Farm           Falafel Restaurant   
3    English Restaurant   Dumpling Restaurant  Eastern European Restaurant   
4          Dessert Shop                Bakery           Seafood Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0  Fast Food Restaurant           Cupcake Shop  
1                 Diner      Indian Restaurant  
2               Factory            Event Space  
3     Electronics Store                    Zoo  
4  Fast Food Restaurant         Ice Cream Shop

Let's create a copy of this dataframe as it will be used for 2 different analysis

In [21]:
neighborhoods_venues_sorted_2=neighborhoods_venues_sorted.copy(deep=True)

## Analysis 1: Cluster various locations on the basis of top venues in the neighborhood

Run *k*-means to cluster the neighborhood into 10 clusters.

In [22]:
# import k-means from clustering stage
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 10

pune_grouped_clustering = pune_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(pune_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([6, 0, 4, 0, 0, 4, 9, 0, 6, 6], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [23]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
neighborhoods_venues_sorted.head()

Cluster Labels    Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0               6    Adarsh Nagar     Indian Restaurant                  Café   
1               0          Akurdi                   Gym                  Café   
2               4          Alandi  Fast Food Restaurant           Bus Station   
3               0     Alandi Road                Bakery           Supermarket   
4               0  Ambedkar Nagar     Indian Restaurant                   Bar   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0    Chinese Restaurant                   Pub                Bakery   
1           Coffee Shop     Convenience Store        Ice Cream Shop   
2     Indian Restaurant                 River            Distillery   
3             Juice Bar           Coffee Shop                 Hotel   
4    Italian Restaurant           Coffee Shop                Lounge   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0           Coffee Shop  Gym / Fitness Center             Asian Restaurant   
1           Snack Place  Fast Food Restaurant    Middle Eastern Restaurant   
2        Farmers Market                  Farm           Falafel Restaurant   
3    English Restaurant   Dumpling Restaurant  Eastern European Restaurant   
4          Dessert Shop                Bakery           Seafood Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0  Fast Food Restaurant           Cupcake Shop  
1                 Diner      Indian Restaurant  
2               Factory            Event Space  
3     Electronics Store                    Zoo  
4  Fast Food Restaurant         Ice Cream Shop

In [24]:
pune_merged = location_df

# merge pune_grouped with location_df to add latitude/longitude for each neighborhood
pune_merged = pune_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Location')
pune_merged=pune_merged[pd.notnull(pune_merged['Cluster Labels'])]

pune_merged.head() # check the last columns!

Location   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0   Bund Garden  18.539848  73.885117             0.0                  Café   
1  Shivajinagar  18.510099  73.817398             6.0     Indian Restaurant   
2         Aundh  18.563162  73.809555             6.0     Indian Restaurant   
3       Kondhwa  18.478436  73.890213             0.0      Asian Restaurant   
4     Chinchwad  18.636131  73.796143             6.0     Indian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Hotel     Indian Restaurant    Italian Restaurant   
1        Ice Cream Shop          Dessert Shop                  Café   
2         Shopping Mall          Dessert Shop  Fast Food Restaurant   
3           Coffee Shop                Bakery            Restaurant   
4         Shopping Mall                 Hotel                   Gym   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery           Coffee Shop  Fast Food Restaurant   
1           Pizza Place           Coffee Shop        Breakfast Spot   
2            Restaurant                Bakery                 Diner   
3            Hookah Bar                  Café            Sports Bar   
4             Multiplex        Sandwich Place  Fast Food Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Chinese Restaurant                Lounge    Japanese Restaurant  
1          Burger Joint  Fast Food Restaurant             Smoke Shop  
2           Snack Place   Sporting Goods Shop         Ice Cream Shop  
3    Mughlai Restaurant             BBQ Joint         Ice Cream Shop  
4             Bookstore           Bus Station             Restaurant

In [25]:
pune_merged['Cluster Labels']=pune_merged['Cluster Labels'].astype('int64')
pune_merged['Cluster Labels'].dtype

dtype('int64')

Finally, let's visualize the resulting clusters

In [26]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
# create map
map_clusters=folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x=np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(pune_merged['Latitude'], pune_merged['Longitude'], pune_merged['Location'], pune_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Analysis Part 2: Distance from nearest 3 Industries

Let's use geopy.distance api to write a method which returns distance between two points in map in KM unit

In [27]:
from geopy import distance
def calculate_distance(point1, point2 ):
    return distance.distance(point1, point2).km

def calculate_for_location(point1, df, j):
    for i in range(industries_df.shape[0]):
        p2=(industries_df.loc[i,'Latitude'],df.loc[i,'Longitude'])
        distance=calculate_distance(point1, p2)
        df.loc[j, industries_df.loc[i,'Industries']]=distance
        
# Usage example
#p1=(pune_merged.loc[2,'Latitude'],pune_merged.loc[2,'Longitude'])
#p2=(pune_merged.loc[1,'Latitude'],pune_merged.loc[1,'Longitude'])
#print("Distance between ",pune_merged.loc[1,'Location'], " and ", pune_merged.loc[2,'Location'], " is : ", calculate_distance(p1, p2))

Let's create another dataframe from location_df having distances from various industrial areas

In [28]:
# create columns according to distance from industries
columns=['Location','Latitude','Longitude']
for i in range(industries_df.shape[0]):
    columns.append( industries_df.loc[i,'Industries'])
distance_from_industries_df=pd.DataFrame(columns=columns)
distance_from_industries_df['Location']=location_df['Location']
distance_from_industries_df['Latitude']=location_df['Latitude']
distance_from_industries_df['Longitude']=location_df['Longitude']
for i in range(distance_from_industries_df.shape[0]):
    p1=(distance_from_industries_df.loc[i,'Latitude'],distance_from_industries_df.loc[i,'Longitude'])
    calculate_for_location(p1,distance_from_industries_df, i)
distance_from_industries_df

Location   Latitude  Longitude  \
0                      Bund Garden  18.539848  73.885117   
1                     Shivajinagar  18.510099  73.817398   
2                            Aundh  18.563162  73.809555   
3                          Kondhwa  18.478436  73.890213   
4                        Chinchwad  18.636131  73.796143   
5                      Satara Road  18.488499  73.857956   
6                          Kothrud  18.508699  73.812500   
7              Senapati Bapat Road  18.534451  73.837349   
8                    Kalyani Nagar  18.548101  73.900070   
9                 Hinjewadi Phase1  18.586555  73.734741   
10                Hinjewadi Phase2  18.598255  73.706207   
11                 Magarpatta City  18.522141  73.931740   
12                      VimanNagar  18.552900  73.892197   
13                           Baner  18.559280  73.790627   
14                       Hinjewadi  18.592600  73.718399   
15                          Kirkee  18.573057  73.879478   
16                    Fatima Nagar  18.503725  73.900597   
17                          Pimpri  18.622713  73.802017   
18    Model Colony - Wealth Branch  18.530758  73.835197   
19                 Pimple Saudagar  18.596691  73.797791   
20                   Sinhagad Road  18.472366  73.821587   
21                      Tilak Road  18.507809  73.844376   
22                         Bavdhan  18.520041  73.778526   
23                          Katraj  18.447878  73.857796   
24           Aundh - Nagardas Road  18.565594  73.806984   
25   Koregaon Park - Wealth Branch  18.538630  73.885826   
26                         Kharadi  18.520477  73.856956   
27                           Nigdi  18.653700  73.770699   
28                   Thermax Chowk  18.660992  73.794052   
29                    Mayur Colony  18.505028  73.817574   
..                             ...        ...        ...   
168                   Sasane Nagar  18.499571  73.933818   
169                   Shanti Nagar  19.064192  72.931151   
170                  Shastri Nagar  19.138039  72.828016   
171                      Shikrapur  18.695989  74.136901   
172                   Shinde Vasti  18.649108  74.193584   
173                  Shivaji Nagar  19.063789  72.924900   
174                        Shivane  18.467041  73.793065   
175                   Siddhi Nagar  18.496901  73.823339   
176                  Someshwarwadi  18.549381  73.798292   
177                     Sopan Baug  18.514171  73.901401   
178                            Sus  18.554159  73.753904   
179                      Sutarwadi  18.538629  73.781421   
180                      Tathawade  18.493309  73.827318   
181                       TherGaon  18.609275  73.772937   
182                    Tukai Nagar  18.473799  73.827719   
183                  Tukaram Nagar  18.614590  73.815697   
184                    Udyog Nagar  19.170000  72.846000   
185                          Undri  18.458157  73.915029   
186                 VadGaon Budruk  18.467497  73.825365   
187                  Vallabh Nagar  19.243475  72.866084   
188                    Viman Nagar  18.520306  73.782938   
189                   Vishal Nagar  18.585322  73.782938   
190                  Vitthal Nagar  18.448006  73.814689   
191                    Vivek Nagar  19.076176  72.905625   
192                 WadGaon Budruk  18.460143  73.826010   
193                  WadGaon Sheri  18.550441  73.917167   
194                          Wakad  18.610902  73.763796   
195                       Wanowrie  18.488368  73.898667   
196                 Yashwant Nagar  19.077216  72.854130   
197                        Yerwada  18.550361  73.879746   

    Pimpri Chinchwad MIDC Rajiv Gandhi InfoTech Park Hinjewadi Phase I  \
0                 9.74938                                      9.16629   
1                 14.8727                                      9.03037   
2                 10.7235                                      3.26381   
3        

In [29]:
distance_from_industries_df.head()

Location   Latitude  Longitude Pimpri Chinchwad MIDC  \
0   Bund Garden  18.539848  73.885117               9.74938   
1  Shivajinagar  18.510099  73.817398               14.8727   
2         Aundh  18.563162  73.809555               10.7235   
3       Kondhwa  18.478436  73.890213               16.5557   
4     Chinchwad  18.636131  73.796143               9.43245   

  Rajiv Gandhi InfoTech Park Hinjewadi Phase I  \
0                                      9.16629   
1                                      9.03037   
2                                      3.26381   
3                                      14.7048   
4                                      5.40698   

  Rajiv Gandhi InfoTech Park Hinjewadi Phase II  \
0                                       10.2672   
1                                       9.79274   
2                                       3.88438   
3                                        15.761   
4                                        4.4249   

  Rajiv Gandhi InfoTech Park Hinjewadi Phase III (SEZ) Magarpatta City  \
0                                            5.77213           9.59648   
1                                            11.8664           2.61032   
2                                            9.08363            4.7562   
3                                            12.5446           11.0493   
4                                            11.0755           12.6173   

  Kharadi Knowledge Park Talawade InfoTech Park Talegaon Floriculture Park  \
0                3.10114                23.4062                    21.5874   
1                 6.1929                25.4146                    24.3746   
2                5.29778                19.5385                    18.6423   
3                8.67514                30.0547                    28.3428   
4                 11.505                11.5884                    11.2105   

  Ranjangaon Industrial Area Chakan Industrial Area  
0                    23.7162                29.2928  
1                    28.3329                29.2475  
2                    23.1451                 23.414  
3                    30.4789                 35.447  
4                    17.0101                15.3953

Let's write a function to sort the distance in ascending order

In [30]:
def return_closest_industries(row, num_top_venues):
    row_categories = row.iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=True)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Let's create a new dataframe with closest 3 industrial areas to each location

In [31]:
import numpy as np
num_top_venues = 3

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Location']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Industry'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Industry'.format(ind+1))

# create a new dataframe
industries_location_sorted= pd.DataFrame(columns=columns)
industries_location_sorted['Location'] = distance_from_industries_df['Location']

for ind in np.arange(distance_from_industries_df.shape[0]):
    industries_location_sorted.iloc[ind, 1:] = return_closest_industries(distance_from_industries_df.iloc[ind, :], num_top_venues)

industries_location_sorted.head()

Location                       1st Most Common Industry  \
0   Bund Garden                         Kharadi Knowledge Park   
1  Shivajinagar                                Magarpatta City   
2         Aundh   Rajiv Gandhi InfoTech Park Hinjewadi Phase I   
3       Kondhwa                         Kharadi Knowledge Park   
4     Chinchwad  Rajiv Gandhi InfoTech Park Hinjewadi Phase II   

                            2nd Most Common Industry  \
0  Rajiv Gandhi InfoTech Park Hinjewadi Phase III...   
1                             Kharadi Knowledge Park   
2      Rajiv Gandhi InfoTech Park Hinjewadi Phase II   
3                                    Magarpatta City   
4       Rajiv Gandhi InfoTech Park Hinjewadi Phase I   

                            3rd Most Common Industry  
0       Rajiv Gandhi InfoTech Park Hinjewadi Phase I  
1       Rajiv Gandhi InfoTech Park Hinjewadi Phase I  
2                                    Magarpatta City  
3  Rajiv Gandhi InfoTech Park Hinjewadi Phase III...  
4                              Pimpri Chinchwad MIDC

Now, let's add this data into our neighborhoods_venues_sorted_2 dataframe for our analysis

In [32]:
neighborhoods_venues_sorted_2['1st Most Common Industry']=industries_location_sorted['1st Most Common Industry']
neighborhoods_venues_sorted_2['2nd Most Common Industry']=industries_location_sorted['2nd Most Common Industry']
neighborhoods_venues_sorted_2['3rd Most Common Industry']=industries_location_sorted['3rd Most Common Industry']

In [33]:
neighborhoods_venues_sorted_2.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0    Adarsh Nagar     Indian Restaurant                  Café   
1          Akurdi                   Gym                  Café   
2          Alandi  Fast Food Restaurant           Bus Station   
3     Alandi Road                Bakery           Supermarket   
4  Ambedkar Nagar     Indian Restaurant                   Bar   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0    Chinese Restaurant                   Pub                Bakery   
1           Coffee Shop     Convenience Store        Ice Cream Shop   
2     Indian Restaurant                 River            Distillery   
3             Juice Bar           Coffee Shop                 Hotel   
4    Italian Restaurant           Coffee Shop                Lounge   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0           Coffee Shop  Gym / Fitness Center             Asian Restaurant   
1           Snack Place  Fast Food Restaurant    Middle Eastern Restaurant   
2        Farmers Market                  Farm           Falafel Restaurant   
3    English Restaurant   Dumpling Restaurant  Eastern European Restaurant   
4          Dessert Shop                Bakery           Seafood Restaurant   

  9th Most Common Venue 10th Most Common Venue  \
0  Fast Food Restaurant           Cupcake Shop   
1                 Diner      Indian Restaurant   
2               Factory            Event Space   
3     Electronics Store                    Zoo   
4  Fast Food Restaurant         Ice Cream Shop   

                        1st Most Common Industry  \
0                         Kharadi Knowledge Park   
1                                Magarpatta City   
2   Rajiv Gandhi InfoTech Park Hinjewadi Phase I   
3                         Kharadi Knowledge Park   
4  Rajiv Gandhi InfoTech Park Hinjewadi Phase II   

                            2nd Most Common Industry  \
0  Rajiv Gandhi InfoTech Park Hinjewadi Phase III...   
1                             Kharadi Knowledge Park   
2      Rajiv Gandhi InfoTech Park Hinjewadi Phase II   
3                                    Magarpatta City   
4       Rajiv Gandhi InfoTech Park Hinjewadi Phase I   

                            3rd Most Common Industry  
0       Rajiv Gandhi InfoTech Park Hinjewadi Phase I  
1       Rajiv Gandhi InfoTech Park Hinjewadi Phase I  
2                                    Magarpatta City  
3  Rajiv Gandhi InfoTech Park Hinjewadi Phase III...  
4                              Pimpri Chinchwad MIDC

Let's create clustering data set on which K-Means clustering, adding all the distance columns in pune_grouped dataset

In [34]:
distance_from_industries_df.columns.shape[0]
for i in range(3, distance_from_industries_df.columns.shape[0]):
    pune_grouped[distance_from_industries_df.columns[i]]=distance_from_industries_df[distance_from_industries_df.columns[i]]

In [35]:
pune_grouped.head()

Neighborhood  ATM  Accessories Store  Adult Boutique  Afghan Restaurant  \
0    Adarsh Nagar  0.0                0.0             0.0                0.0   
1          Akurdi  0.0                0.0             0.0                0.0   
2          Alandi  0.0                0.0             0.0                0.0   
3     Alandi Road  0.0                0.0             0.0                0.0   
4  Ambedkar Nagar  0.0                0.0             0.0                0.0   

   Airport  Airport Lounge  American Restaurant  Andhra Restaurant  Arcade  \
0      0.0             0.0                 0.01                0.0    0.00   
1      0.0             0.0                 0.00                0.0    0.00   
2      0.0             0.0                 0.00                0.0    0.00   
3      0.0             0.0                 0.00                0.0    0.00   
4      0.0             0.0                 0.01                0.0    0.02   

            ...            Pimpri Chinchwad MIDC  \
0           ...                          9.74938   
1           ...                          14.8727   
2           ...                          10.7235   
3           ...                          16.5557   
4           ...                          9.43245   

   Rajiv Gandhi InfoTech Park Hinjewadi Phase I  \
0                                       9.16629   
1                                       9.03037   
2                                       3.26381   
3                                       14.7048   
4                                       5.40698   

   Rajiv Gandhi InfoTech Park Hinjewadi Phase II  \
0                                        10.2672   
1                                        9.79274   
2                                        3.88438   
3                                         15.761   
4                                         4.4249   

   Rajiv Gandhi InfoTech Park Hinjewadi Phase III (SEZ)  Magarpatta City  \
0                                            5.77213             9.59648   
1                                            11.8664             2.61032   
2                                            9.08363              4.7562   
3                                            12.5446             11.0493   
4                                            11.0755             12.6173   

   Kharadi Knowledge Park  Talawade InfoTech Park  Talegaon Floriculture Park  \
0                 3.10114                 23.4062                     21.5874   
1                  6.1929                 25.4146                     24.3746   
2                 5.29778                 19.5385                     18.6423   
3                 8.67514                 30.0547                     28.3428   
4                  11.505                 11.5884                     11.2105   

   Ranjangaon Industrial Area  Chakan Industrial Area  
0                     23.7162                 29.2928  
1                     28.3329                 29.2475  
2                     23.1451                  23.414  
3                     30.4789                  35.447  
4                     17.0101                 15.3953  

[5 rows x 244 columns]

The values of distance in last 10 columns in much greater than remaining columns, here we will perform feature scaling to get an uniformly distributed feature values

In [37]:
from sklearn import preprocessing
pune_grouped=pune_grouped.drop('Neighborhood',1)
pune_grouped = preprocessing.StandardScaler().fit(pune_grouped).transform(pune_grouped)

/home/priyesh/anaconda3/lib/python3.7/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype float64, object were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/home/priyesh/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: DataConversionWarning: Data with input dtype float64, object were all converted to float64 by StandardScaler.
  This is separate from the ipykernel package so we can avoid doing imports until


In [38]:
pune_grouped

array([[-0.14720738, -0.17991822, -0.07312724, ..., -0.41292813,
        -0.4398507 , -0.33642179],
       [-0.14720738, -0.17991822, -0.07312724, ..., -0.34331094,
        -0.32433824, -0.3376475 ],
       [-0.14720738, -0.17991822, -0.07312724, ..., -0.48648841,
        -0.4541385 , -0.49535871],
       ...,
       [-0.14720738, -0.17991822, -0.07312724, ..., -0.17453601,
        -0.21397552, -0.08289283],
       [-0.14720738, -0.17991822, -0.07312724, ..., -0.2271175 ,
        -0.21661077, -0.20903189],
       [-0.14720738, -0.17991822, -0.07312724, ...,  1.97107405,
         2.00990175,  1.73301568]])

Run *k*-means to cluster the neighborhood into 10 clusters. 

In [39]:
# import k-means from clustering stage
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 10

#pune_grouped_clustering = pune_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(pune_grouped)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 3, 1, 1, 1, 4, 4], dtype=int32)

add clustering labels

In [40]:
# add clustering labels
neighborhoods_venues_sorted_2.insert(0, 'Cluster Labels', kmeans.labels_)
neighborhoods_venues_sorted_2.head()

Cluster Labels    Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0               1    Adarsh Nagar     Indian Restaurant                  Café   
1               1          Akurdi                   Gym                  Café   
2               1          Alandi  Fast Food Restaurant           Bus Station   
3               1     Alandi Road                Bakery           Supermarket   
4               3  Ambedkar Nagar     Indian Restaurant                   Bar   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0    Chinese Restaurant                   Pub                Bakery   
1           Coffee Shop     Convenience Store        Ice Cream Shop   
2     Indian Restaurant                 River            Distillery   
3             Juice Bar           Coffee Shop                 Hotel   
4    Italian Restaurant           Coffee Shop                Lounge   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0           Coffee Shop  Gym / Fitness Center             Asian Restaurant   
1           Snack Place  Fast Food Restaurant    Middle Eastern Restaurant   
2        Farmers Market                  Farm           Falafel Restaurant   
3    English Restaurant   Dumpling Restaurant  Eastern European Restaurant   
4          Dessert Shop                Bakery           Seafood Restaurant   

  9th Most Common Venue 10th Most Common Venue  \
0  Fast Food Restaurant           Cupcake Shop   
1                 Diner      Indian Restaurant   
2               Factory            Event Space   
3     Electronics Store                    Zoo   
4  Fast Food Restaurant         Ice Cream Shop   

                        1st Most Common Industry  \
0                         Kharadi Knowledge Park   
1                                Magarpatta City   
2   Rajiv Gandhi InfoTech Park Hinjewadi Phase I   
3                         Kharadi Knowledge Park   
4  Rajiv Gandhi InfoTech Park Hinjewadi Phase II   

                            2nd Most Common Industry  \
0  Rajiv Gandhi InfoTech Park Hinjewadi Phase III...   
1                             Kharadi Knowledge Park   
2      Rajiv Gandhi InfoTech Park Hinjewadi Phase II   
3                                    Magarpatta City   
4       Rajiv Gandhi InfoTech Park Hinjewadi Phase I   

                            3rd Most Common Industry  
0       Rajiv Gandhi InfoTech Park Hinjewadi Phase I  
1       Rajiv Gandhi InfoTech Park Hinjewadi Phase I  
2                                    Magarpatta City  
3  Rajiv Gandhi InfoTech Park Hinjewadi Phase III...  
4                              Pimpri Chinchwad MIDC

In [41]:
pune_merged_2 = location_df

# merge pune_grouped with location_df to add latitude/longitude for each neighborhood
pune_merged_2 = pune_merged_2.join(neighborhoods_venues_sorted_2.set_index('Neighborhood'), on='Location')
pune_merged_2=pune_merged_2[pd.notnull(pune_merged_2['Cluster Labels'])]

pune_merged_2.head() # check the last columns!

Location   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0   Bund Garden  18.539848  73.885117             1.0                  Café   
1  Shivajinagar  18.510099  73.817398             1.0     Indian Restaurant   
2         Aundh  18.563162  73.809555             4.0     Indian Restaurant   
3       Kondhwa  18.478436  73.890213             1.0      Asian Restaurant   
4     Chinchwad  18.636131  73.796143             7.0     Indian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Hotel     Indian Restaurant    Italian Restaurant   
1        Ice Cream Shop          Dessert Shop                  Café   
2         Shopping Mall          Dessert Shop  Fast Food Restaurant   
3           Coffee Shop                Bakery            Restaurant   
4         Shopping Mall                 Hotel                   Gym   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery           Coffee Shop  Fast Food Restaurant   
1           Pizza Place           Coffee Shop        Breakfast Spot   
2            Restaurant                Bakery                 Diner   
3            Hookah Bar                  Café            Sports Bar   
4             Multiplex        Sandwich Place  Fast Food Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
0    Chinese Restaurant                Lounge    Japanese Restaurant   
1          Burger Joint  Fast Food Restaurant             Smoke Shop   
2           Snack Place   Sporting Goods Shop         Ice Cream Shop   
3    Mughlai Restaurant             BBQ Joint         Ice Cream Shop   
4             Bookstore           Bus Station             Restaurant   

                            1st Most Common Industry  \
0  Rajiv Gandhi InfoTech Park Hinjewadi Phase III...   
1                                    Magarpatta City   
2                             Kharadi Knowledge Park   
3                              Pimpri Chinchwad MIDC   
4                             Kharadi Knowledge Park   

                            2nd Most Common Industry  \
0                             Kharadi Knowledge Park   
1       Rajiv Gandhi InfoTech Park Hinjewadi Phase I   
2  Rajiv Gandhi InfoTech Park Hinjewadi Phase III...   
3                         Ranjangaon Industrial Area   
4  Rajiv Gandhi InfoTech Park Hinjewadi Phase III...   

                        3rd Most Common Industry  
0                          Pimpri Chinchwad MIDC  
1  Rajiv Gandhi InfoTech Park Hinjewadi Phase II  
2                          Pimpri Chinchwad MIDC  
3                     Talegaon Floriculture Park  
4                                Magarpatta City

In [42]:
pune_merged_2['Cluster Labels']=pune_merged_2['Cluster Labels'].astype('int64')
pune_merged_2['Cluster Labels'].dtype

dtype('int64')

Let's visualize resulting clusters

In [43]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
# create map
map_clusters=folium.Map(location=[latitude, longitude], zoom_start=11)
# set color scheme for the clusters
x=np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(pune_merged_2['Latitude'], pune_merged_2['Longitude'], pune_merged_2['Location'], pune_merged_2['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

The identification of these clusters is more clear and widely distributed across the city. Now let's analyse a few cluster and try to identify difference between them.

## Analysis <a name="analysis"></a>

#### Analysis of Part 1

Let us examine and compare any 3 clusters of Part 1 analysis with 3 clusters of Part 2 analysis clusters.

#### Cluster 1 of Part 1 analysis

In [60]:
pune_merged.loc[pune_merged['Cluster Labels'] == 4, pune_merged.columns[[0] + list(range(4, pune_merged.shape[1]))]]

Location 1st Most Common Venue  \
20          Sinhagad Road     Indian Restaurant   
33                  Warje           Coffee Shop   
53              Paud Road     Indian Restaurant   
67               Fursungi  Fast Food Restaurant   
76   Prabhadevi Tech Park     Indian Restaurant   
81                 Alandi  Fast Food Restaurant   
84            Anand Nagar  Fast Food Restaurant   
90         Balewadi Phata     Indian Restaurant   
95         Bhusari Colony     Indian Restaurant   
111        Digambar Nagar     Indian Restaurant   
119          Hingne Khurd           Snack Place   
123            Karve Road     Indian Restaurant   
124             Kasarwadi     Indian Restaurant   
150            Panchawati                  Park   
182           Tukai Nagar     Indian Restaurant   
183         Tukaram Nagar     Indian Restaurant   
184           Udyog Nagar     Indian Restaurant   
186        VadGaon Budruk  Fast Food Restaurant   
194                 Wakad  Fast Food Restaurant   
197               Yerwada             Gastropub   

             2nd Most Common Venue      3rd Most Common Venue  \
20                           Diner             Ice Cream Shop   
33               Indian Restaurant                Pizza Place   
53                            Café        Sporting Goods Shop   
67                           Hotel          Indian Restaurant   
76            Fast Food Restaurant                Snack Place   
81                     Bus Station          Indian Restaurant   
84                          Bakery          Indian Restaurant   
90            Fast Food Restaurant             Breakfast Spot   
95                            Café       Fast Food Restaurant   
111           Fast Food Restaurant                     Bakery   
119                    Pizza Place       Fast Food Restaurant   
123           Fast Food Restaurant                      Diner   
124           Fast Food Restaurant  Indian Chinese Restaurant   
150                        Airport            Motorcycle Shop   
182           Fast Food Restaurant             Ice Cream Shop   
183                    Bus Station             Cosmetics Shop   
184           Fast Food Restaurant                Snack Place   
186                           Café          Indian Restaurant   
194  Vegetarian / Vegan Restaurant          Indian Restaurant   
197              Indian Restaurant                        Bar   

    4th Most Common Venue 5th Most Common Venue  \
20                   Café                Bakery   
33   Fast Food Restaurant         Grocery Store   
53        Motorcycle Shop        Breakfast Spot   
67              Rock Club            Distillery   
76       Asian Restaurant         Shopping Mall   
81                  River            Distillery   
84            Coffee Shop           Snack Place   
90                   Café                Lounge   
95            Bus Station        Breakfast Spot   
111                 Hotel                  Farm   
119  Gym / Fitness Center     Indian Restaurant   
123           Snack Place           Coffee Shop   
124    Italian Restaurant              Platform   
150     Indian Restaurant                   Zoo   
182                Bakery                 Diner   
183             Hotel Bar           Snack Place   
184                Bakery              Pharmacy   
186                 Plaza                 Diner   
194          Dessert Shop        Ice Cream Shop   
197  Fast Food Restaurant            Playground   

             6th Most Common Venue        7th Most Common Venue  \
20            Fast Food Restaurant         Gym / Fitness Center   
33                           Diner  Eastern European Restaurant   
53                     Bus Station                        Diner   
67                  Farmers Market                         Farm   
76                          Lounge      South Indian Restaurant   
81                  Farmers Market                         Farm   
84                  Ice Cream Shop    

pune_merged.loc[pune_merged['Cluster Labels'] == 0, pune_merged.columns[[1] + list(range(5, pune_merged.shape[1]))]]

#### Cluster 2 of Part 1 analysis

In [50]:
pune_merged.loc[pune_merged['Cluster Labels'] == 1, pune_merged.columns[[0] + list(range(4, pune_merged.shape[1]))]]

Location 1st Most Common Venue 2nd Most Common Venue  \
35          Karanjepul     Mobile Phone Shop                   Zoo   
40  Pimpalgaon - Daund     Mobile Phone Shop                   Zoo   
41             Pargaon     Mobile Phone Shop                   Zoo   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
35            Donut Shop  Fast Food Restaurant        Farmers Market   
40            Donut Shop  Fast Food Restaurant        Farmers Market   
41            Donut Shop  Fast Food Restaurant        Farmers Market   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
35                  Farm    Falafel Restaurant               Factory   
40                  Farm    Falafel Restaurant               Factory   
41                  Farm    Falafel Restaurant               Factory   

   9th Most Common Venue 10th Most Common Venue  
35           Event Space     English Restaurant  
40           Event Space     English Restaurant  
41           Event Space     English Restaurant

#### Cluster 3 of Part 1 analysis

In [53]:
pune_merged.loc[pune_merged['Cluster Labels'] == 3, pune_merged.columns[[0] + list(range(4, pune_merged.shape[1]))]]

Location 1st Most Common Venue 2nd Most Common Venue  \
46                 Bhigwan     Indian Restaurant         Train Station   
69   Manjari Road Hadapsar     Indian Restaurant    Seafood Restaurant   
106              Dattawadi     Indian Restaurant                 Trail   
114                Gadital     Indian Restaurant  Fast Food Restaurant   
151                 Pashan     Indian Restaurant        Farmers Market   
178                    Sus     Indian Restaurant                Resort   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
46     Seafood Restaurant                   Zoo            Distillery   
69                    Zoo            Distillery        Farmers Market   
106           Pizza Place            Food Truck     Electronics Store   
114        Farmers Market           Coffee Shop      Department Store   
151         Shopping Mall    Seafood Restaurant            Distillery   
178                   Zoo                 Diner        Farmers Market   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
46         Farmers Market                  Farm    Falafel Restaurant   
69                   Farm    Falafel Restaurant               Factory   
106            Donut Shop            Dosa Place   Dumpling Restaurant   
114                Bakery                   Zoo            Dosa Place   
151                  Farm    Falafel Restaurant               Factory   
178                  Farm    Falafel Restaurant               Factory   

           9th Most Common Venue 10th Most Common Venue  
46                       Factory            Event Space  
69                   Event Space     English Restaurant  
106  Eastern European Restaurant                    Zoo  
114                         Farm     Falafel Restaurant  
151                  Event Space     English Restaurant  
178                  Event Space     English Restaurant

#### Analysis of Part 2

#### Cluster 1 of Part 2 analysis

In [54]:
pune_merged_2.loc[pune_merged_2['Cluster Labels'] == 0, pune_merged_2.columns[[0] + list(range(4, pune_merged_2.shape[1]))]]

Location 1st Most Common Venue            2nd Most Common Venue  \
5        Satara Road     Indian Restaurant                           Bakery   
47     Sahakar Nagar     Indian Restaurant  Southern / Soul Food Restaurant   
62      Bhosle Nagar     Indian Restaurant                      Coffee Shop   
152  Pashan-Sus Road        Breakfast Spot    Vegetarian / Vegan Restaurant   

               3rd Most Common Venue          4th Most Common Venue  \
5                     Ice Cream Shop  Vegetarian / Vegan Restaurant   
47                       Coffee Shop           Gym / Fitness Center   
62   Southern / Soul Food Restaurant           Fast Food Restaurant   
152                Indian Restaurant                    Coffee Shop   

               5th Most Common Venue 6th Most Common Venue  \
5    Southern / Soul Food Restaurant           Coffee Shop   
47                    Ice Cream Shop         Shopping Mall   
62                  Asian Restaurant                Garden   
152                         Mountain    Italian Restaurant   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
5          Breakfast Spot         Shopping Mall  Fast Food Restaurant   
47         Breakfast Spot     Electronics Store           Bus Station   
62             Shoe Store          Tennis Court             Multiplex   
152                 Diner           Beer Garden        Ice Cream Shop   

    10th Most Common Venue                       1st Most Common Industry  \
5                   Bistro                         Kharadi Knowledge Park   
47                  Bistro  Rajiv Gandhi InfoTech Park Hinjewadi Phase II   
62          Ice Cream Shop                         Kharadi Knowledge Park   
152             Food Court                         Kharadi Knowledge Park   

                              2nd Most Common Industry  \
5                                      Magarpatta City   
47        Rajiv Gandhi InfoTech Park Hinjewadi Phase I   
62                                     Magarpatta City   
152  Rajiv Gandhi InfoTech Park Hinjewadi Phase III...   

                              3rd Most Common Industry  
5    Rajiv Gandhi InfoTech Park Hinjewadi Phase III...  
47                                     Magarpatta City  
62   Rajiv Gandhi InfoTech Park Hinjewadi Phase III...  
152                                    Magarpatta City

#### Cluster 2 of Part 2 analysis

In [59]:
pune_merged_2.loc[pune_merged_2['Cluster Labels'] == 5, pune_merged_2.columns[[0] + list(range(4, pune_merged_2.shape[1]))]]

Location 1st Most Common Venue    2nd Most Common Venue  \
31     WTC-Kharadi           Coffee Shop              Pizza Place   
112  Eon Free Zone     Indian Restaurant  North Indian Restaurant   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
31      Indian Restaurant            Irani Cafe                  Café   
112           Coffee Shop  Fast Food Restaurant             Cafeteria   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
31              Cafeteria                   Pub         Go Kart Track   
112                  Café                   Pub         Go Kart Track   

       9th Most Common Venue 10th Most Common Venue 1st Most Common Industry  \
31   North Indian Restaurant                  Diner          Magarpatta City   
112               Irani Cafe            Pizza Place   Kharadi Knowledge Park   

                         2nd Most Common Industry  \
31   Rajiv Gandhi InfoTech Park Hinjewadi Phase I   
112                               Magarpatta City   

                              3rd Most Common Industry  
31       Rajiv Gandhi InfoTech Park Hinjewadi Phase II  
112  Rajiv Gandhi InfoTech Park Hinjewadi Phase III...

#### Cluster 3 of Part 2 analysis

In [58]:
pune_merged_2.loc[pune_merged_2['Cluster Labels'] == 4, pune_merged_2.columns[[0] + list(range(4, pune_merged_2.shape[1]))]]

Location 1st Most Common Venue 2nd Most Common Venue  \
2                    Aundh     Indian Restaurant         Shopping Mall   
24   Aundh - Nagardas Road     Indian Restaurant  Fast Food Restaurant   
87            Aundh Annexe     Indian Restaurant          Dessert Shop   
88              Aundh Gaon     Indian Restaurant          Dessert Shop   
134            Kunj Colony     Indian Restaurant                  Café   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2            Dessert Shop  Fast Food Restaurant            Restaurant   
24          Shopping Mall   Sporting Goods Shop          Dessert Shop   
87   Fast Food Restaurant         Shopping Mall            Restaurant   
88          Shopping Mall            Restaurant  Fast Food Restaurant   
134                   Bar  Fast Food Restaurant        Ice Cream Shop   

    6th Most Common Venue 7th Most Common Venue  \
2                  Bakery                 Diner   
24             Restaurant           Snack Place   
87                   Café        Sandwich Place   
88            Coffee Shop           Snack Place   
134           Snack Place           Coffee Shop   

             8th Most Common Venue    9th Most Common Venue  \
2                      Snack Place      Sporting Goods Shop   
24                          Bakery            Grocery Store   
87                          Bakery      Sporting Goods Shop   
88             Sporting Goods Shop           Ice Cream Shop   
134  Vegetarian / Vegan Restaurant  South Indian Restaurant   

    10th Most Common Venue                       1st Most Common Industry  \
2           Ice Cream Shop                         Kharadi Knowledge Park   
24                   Diner  Rajiv Gandhi InfoTech Park Hinjewadi Phase II   
87           Grocery Store  Rajiv Gandhi InfoTech Park Hinjewadi Phase II   
88                  Bakery                         Kharadi Knowledge Park   
134              Juice Bar   Rajiv Gandhi InfoTech Park Hinjewadi Phase I   

                              2nd Most Common Industry  \
2    Rajiv Gandhi InfoTech Park Hinjewadi Phase III...   
24        Rajiv Gandhi InfoTech Park Hinjewadi Phase I   
87        Rajiv Gandhi InfoTech Park Hinjewadi Phase I   
88   Rajiv Gandhi InfoTech Park Hinjewadi Phase III...   
134      Rajiv Gandhi InfoTech Park Hinjewadi Phase II   

    3rd Most Common Industry  
2      Pimpri Chinchwad MIDC  
24           Magarpatta City  
87           Magarpatta City  
88     Pimpri Chinchwad MIDC  
134          Magarpatta City

Now, These clusters have many areas in common, Some of the most common venues of each analysis is as follows: -
-  Analysis 1
    -  Cluster 1: - 
        Indian Restaurants, Bakery, Shopping Mall, Gym 
    -  Cluster 2: -
        Mobile Shops, Zoo, Donut Shop, Fast Food Restaurant 
    -  Cluster 3: -
        Indian Restaurant, Fast Food Restaurant, Seafood Restaurant, Resort 
-  Analysis 3
    -  Cluster 1: -
        Indian Restaurant, Coffee Shop, Icecream Shop and close to Kharadi Knowledge Park and Magarpatta city
    -  Cluster 2: -
        Coffee Shop, Indian Restaurant, Irani Cafe and close to Magarpatta city and Hinjewadi Phase 1
    -  Cluster 3: -
        Indian Restaurant, Dessert Shop, Shopping Mall and close to Rajiv Gandhi Infotech Park and Kharadi Knowledge Park.
        

## Results and Discussion <a name="results"></a>

Our analysis shows that although their are various locations in Pune city. They can be classified into common clusters  according to the venues around 1 K.M area of those clusters. 
First we used all the locations in the Pune city and clustered them according to the frequency of venues around them and classified these 197 locations into 10 clusters.

In our second analysis, we choose 10 industrial locations and included them in our analysis also, to check for any difference in the clusters formed.

We found that the locations are classified differently in both clusters, the distance of each industrial location is also taken into consideration and then classified into clusters.

Although, the clusters are unevenly distributed in each cases, hence grouping most of the locations in 3-4 clusters only, which proves that Pune city has almost all similar set of venues available in each and every geography of the city. If a person/stakeholder tries to locate a place similar to another place at a far distant location, the probability of finding such an location is higher in our case.

## Conclusion <a name="conclusion"></a>

The objective of this analysis was to classify similar locations of Pune city India. We have used two approaches to classify location across this city.

Although, the final decision of selecting a particular location will be made by our stakeholder based on specific characteristics of neighborhoods and locations in every recommended zone, taking into consideration additional factors like attractiveness of each location (proximity to park or water), distance from work location, traffic on roads, levels of noise / proximity to major roads, real estate availability, prices, social and economic dynamics of every neighborhood etc.